spring/boot2-autoconfigure/src/main/java/com/linecorp/armeria/spring/ArmeriaServerGracefulShutdownLifecycle.java

```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2020 LINE Corporation
+ *
+ * LINE Corporation licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package com.linecorp.armeria.spring;
+
+import org.springframework.context.SmartLifecycle;
+
+import com.linecorp.armeria.server.Server;
+
+/**
+ * Make Armeria Server utilize spring's SmartLifecycle feature.
+ * So Armeria will shutdown before other web servers and beans in the context.
+ */
+public class ArmeriaServerGracefulShutdownLifecycle implements SmartLifecycle {
+    /**
+     * {@link Server} created by {@link ArmeriaAutoConfiguration}. .
+     */
+    private final Server server;
+
+    /**
+     * Creates a new instance.
+     */
+    public ArmeriaServerGracefulShutdownLifecycle(Server server) {
+        this.server = server;
+    }
+
+    /**
+     * Return the created {@link Server}.
+     */
+    public Server getServer() {
+        return server;
+    }
+
+    /**
+     * Start this component.
+     * Currently AbstractArmeriaAutoConfiguration help starting the server.
+     */
+    @Override
+    public void start() {
+    }
+
+    /**
+     * Stop this component. This class implements {@link SmartLifecycle}, so don't need to support sync stop.
+     */
+    @Override
+    public void stop() {
+        throw new UnsupportedOperationException("Stop must not be invoked directly");
+    }
+
+    /**
+     * Stop this component.
+     */
+    @Override
+    public void stop(Runnable callback) {
+        server.stop().whenComplete((unused, throwable) -> callback.run());
```

Comment:
This is not called frequently. JFYI, (AFAIK) the performance of `whenComplete()` in Java8 is poor than `handle()`.

Reply:
I see, due to not frequently, maybe let me keep `whenComplete` for just calling callback without returning/rethrow.